### ISSUE TITLE
Integrate GTPO (Group-relative Trajectory-based Policy Optimization) as a new loss_type in GRPOTrainer

### ISSUE THREAD

**Issue by winstonsmith1897** [huggingface/trl#5226]
## Motivation
Hi TRL Team! This issue proposes the integration of Group-relative Trajectory-based Policy Optimization (GTPO) into TRL's `GRPOTrainer`. GTPO addresses fundamental structural vulnerabilities in vanilla GRPO, namely, training instability (policy collapse) and gradient conflicts on shared tokens, enabling highly efficient, KL-free reasoning post-training. As one of the authors of the paper, I would love to volunteer to implement it and submit a PR if the team is open to this addition.

<img width="304" height="280" alt="Image" src="https://github.com/user-attachments/assets/382dd83b-c3c6-4b3c-9222-64e3cdb6f99d" />

## Core Mechanisms of GTPO

<img width="638" height="341" alt="Image" src="https://github.com/user-attachments/assets/20639f5e-82bf-4460-b479-9c9ddf73e9a9" />

GTPO stabilizes training dynamics through two rigorous architectural corrections:

1. **Conflict-Aware Gradient Correction:** Vanilla GRPO systematically applies contradictory gradient updates to shared contiguous prefix/suffix tokens (such as structural and critical formatting tags). GTPO mathematically resolves this by computing binary masks that selectively zero-out negative updates on these conflict regions while amplifying positive ones. This preserves structural consistency without distorting the semantic generation in the unmasked trajectory center.
2. **Entropy Control:** GRPO's zero-mean advantage scaling heavily penalizes confident responses when negatively rewarded. This occurs especially during training stages where the reward is growing and token probability distributions become highly peaked, meaning the model is highly confident in its generation. This disproportionate penalization gradually flattens the token likelihood distribution and shifts model decisions toward unlikely tokens. GTPO mitigates this "policy collapse" through a strict completion filter, discarding trajectories exceeding a provable Shannon entropy threshold, combined with an entropy regularization penalty. This penalty is also crucial for controlling the lower bound of the average entropy reached during training, ensuring that exploration is not prematurely suppressed.

<p align="center"> $$ \mathcal{J}_{\text{GTPO}} = \mathbb{E}_{q,\{o_i\}} \left[\frac{1}{G} \sum_{i=1}^{G} \frac{\delta_i \cdot (\mathcal{A}_i - \gamma \cdot \langle H \rangle_i ) }{|o_i|} \sum_{t = 1}^{|o_i|} r_{i,t}\lambda_{i,t}\right] $$ </p> <p align="center"> $$ \delta_i = \begin{cases} 1, & \text{if } \langle H \rangle_{ini} > \ln 2, \\ 0, & \text{if } \langle H \rangle_{ini} < \ln 2 \text{ and } \langle H \rangle_i > \ln 2, \\ 1, & \text{if } \langle H \rangle_{ini} < \ln 2 \text{ and } \langle H \rangle_i \leq \ln 2. \end{cases} $$ </p>

<img width="623" height="245" alt="Image" src="https://github.com/user-attachments/assets/33527a4b-1b93-4739-a588-d15180c2c4bc" />

## Theoretical Backing vs. Existing GRPO Variants
In our recent theoretical unification of group-based RL methods  (https://arxiv.org/abs/2601.05002), we demonstrate that GRPO-style algorithms (including recent additions like DAPO, GSPO, and CPPO) inherits the same hidden objective biases. Specifically, non-uniform weighting across these methods inevitably induces systematic gradient biases on shared prefix tokens. Furthermore, optimizer momentum (e.g., AdamW) can push policy updates beyond intended clipping regions during repeated inner steps, compounding instability.

Because these variants suffer from the exact same foundational token-level conflicts and momentum-induced drifts as vanilla GRPO, they implicitly rely on the computationally expensive KL-divergence penalty to anchor the policy and prevent collapse. GTPO, by contrast, structurally neutralizes these vulnerabilities at the token and trajectory levels. By natively solving gradient conflicts and entropy explosion, GTPO is uniquely equipped to safely eliminate the KL penalty entirely, maximizing memory efficiency without compromising optimization stability.

## Alignment with TRL's Architecture (Efficiency & Stability)
* **Peak VRAM Efficiency:** By structurally obviating the need for KL-divergence, GTPO requires no reference model during training, strictly reducing VRAM overhead and speeding up the pipeline.
* **Throughput Stability:** Policy collapse in standard GRPO is strongly correlated with entropy explosion, which inherently triggers unbounded `completion_length` growth. GTPO's strict entropy constraints keep generation lengths predictable, averting the severe compute bottlenecks common in late-stage GRPO rollouts.

![Image](https://github.com/user-attachments/assets/25f37b7d-b66e-44fe-88e6-daf8fdae30f7)

## Proposed Implementation
To maintain architectural consistency with recent additions, I propose integrating this logic directly into `GRPOTrainer` via the `GRPOConfig`. 
By setting `loss_type="gtpo"`, the trainer would conditionally apply the conflict masks and entropy filters during the advantage/loss computation step. Relevant hyperparameters (e.g., `gtpo_entropy_threshold`, `gtpo_gamma`) would simply be added to the configuration.

I am available to open a PR and implement this feature. I welcome feedback from the maintainers on this proposed integration path.

### Open source status

- [x] The method implementation is available
- [x] The model weights are available
- [x] The training datasets are available

## Required Information
* **Method Description & Papers:** * *GTPO: Stabilizing Group Relative Policy Optimization via Gradient and Entropy Control* (https://arxiv.org/abs/2508.03772) 
  * *On the Hidden Objective Biases of Group-based Reinforcement Learning* (https://arxiv.org/abs/2601.05002) 
* **Open-Source Implementation:** The official codebase is available at https://github.com/winstonsmith1897/GTPO.
* **Model Weights:** Training scripts and configurations for LLaMA 8B and Qwen 2.5 3B are provided in the authors' repository.

**Comment by qgallouedec** [comment]
Hey @winstonsmith1897, apologies for the slow reply, we're catching up on the backlog.

Thanks for the thorough write-up, and for offering to implement it. GTPO fits the shape of what `GRPOTrainer` already supports we host a number of `loss_type` variants alongside the base objective, so it's not out of scope.

The hesitation is signal and surface: it's so far a single-requester thread, the method has modest external traction (~11 citations, no adoption in the usual RL libraries yet), and it would add new config-field contracts (`gtpo_entropy_threshold`, `gtpo_gamma`) plus trajectory-filtering and token-masking logic that are hard to deprecate once shipped.

Rather than close it, let's keep it open to gather interest, if others want this, please 👍 or comment with your use case.

**Comment by albertvillanova** [comment]
Hey @winstonsmith1897, following up on my note from July: we left this open to gather interest and asked for 👍/use cases, and six weeks on the thread is still at one requester and zero reactions — so I'm closing it for the core `GRPOTrainer` path rather than leaving it open indefinitely.

Two things worth flagging if you still want to land it. First, `loss_type` in TRL is narrowly the per-token loss form plus the normalization strategy (https://github.com/huggingface/trl/blob/b3b5d4b5e9290177ab14b1400ca9590b57b8b0c8/trl/trainer/grpo_trainer.py#L3171-L3200), whereas GTPO's δ trajectory filter and conflict masks act on the advantages and on a new token mask — so it wouldn't fit that slot cleanly, and the integration would be wider than a new enum value. Second, entropy regularization with adaptive control landed in #6140 just after my last reply (https://github.com/huggingface/trl/blob/b3b5d4b5e9290177ab14b1400ca9590b57b8b0c8/trl/trainer/grpo_config.py#L298-L325), so part of GTPO's entropy component is already available to try today.

Your offer to implement is still genuinely welcome — just as `trl.experimental.gtpo` rather than in the stable trainer. That's the documented path for new methods (https://github.com/huggingface/trl/blob/b3b5d4b5e9290177ab14b1400ca9590b57b8b0c8/docs/source/experimental_overview.md), it sits alongside `gmpo`, `gspo_token` and `a2po`, and it lets people actually use it without locking us into config-field contracts before there's adoption. Happy to revisit core integration once there's usage signal.
